homeassistant/components/recorder/purge.py

```diff
@@ -86,52 +88,80 @@ def purge_old_data(
-    return [event.event_id for event in events]
-
-
-def _select_state_and_attributes_ids_to_purge(
-    session: Session, purge_before: datetime, event_ids: list[int]
-) -> tuple[set[int], set[int]]:
-    """Return a list of state ids to purge."""
-    if not event_ids:
-        return set(), set()
-    states = (
-        session.query(States.state_id, States.attributes_id)
-        .filter(States.last_updated < purge_before)
-        .filter(States.event_id.in_(event_ids))
-        .all()
-    )
-    _LOGGER.debug("Selected %s state ids to remove", len(states))
+    event_ids = set()
     state_ids = set()
     attributes_ids = set()
-    for state in states:
-        state_ids.add(state.state_id)
-        if state.attributes_id:
-            attributes_ids.add(state.attributes_id)
-    return state_ids, attributes_ids
+    for event in events:
+        event_ids.add(event.event_id)
+        if event.state_id:
+            state_ids.add(event.state_id)
+        if event.attributes_id:
+            attributes_ids.add(event.attributes_id)
+    return event_ids, state_ids, attributes_ids
 
 
 def _select_unused_attributes_ids(
-    session: Session, attributes_ids: set[int]
+    session: Session, attributes_ids: set[int], using_sqlite: bool
 ) -> set[int]:
     """Return a set of attributes ids that are not used by any states in the database."""
     if not attributes_ids:
         return set()
-    to_remove = attributes_ids - {
-        state[0]
-        for state in session.query(distinct(States.attributes_id))
-        .filter(States.attributes_id.in_(attributes_ids))
-        .all()
-    }
+
+    if using_sqlite:
+        id_query = session.query(distinct(States.attributes_id)).filter(
+            States.attributes_id.in_(attributes_ids)
+        )
+    else:
+        #
+        # We are jumping through hoops a bit here to handle
+        # a case where sqlite performance on finding unused attributes_id
+        # far exceeds other DBMS
+        #
+        # MariaDB/MySQL cannot optimize that query well and has to examine
+        # all the rows that match
+        #
+        # > explain select distinct attributes_id from states where attributes_id in (136723);
+        # +------+-------------+--------+------+-------------------------+-------------------------+---------+-------+-------+-------------+
+        # | id   | select_type | table  | type | possible_keys           | key                     | key_len | ref   | rows  | Extra       |
+        # +------+-------------+--------+------+-------------------------+-------------------------+---------+-------+-------+-------------+
+        # |    1 | SIMPLE      | states | ref  | ix_states_attributes_id | ix_states_attributes_id | 5       | const | 22842 | Using index |
+        # +------+-------------+--------+------+-------------------------+-------------------------+---------+-------+-------+-------------+
+        #
+        # With a single query, it can be optimized away
```

Comment:
What is "it" in the sentence below, can it be rephrased to be clearer?

```suggestion
        # With a single query, it can be optimized away:
```

Reply:
Removed `it`